### PR TITLE
mgr: load modules in separate python sub-interpreters

### PR DESCRIPTION
**Pull request by tserong**
This provides a reasonable amount of isolation between mgr
modules.  Notably, with this change, it's possible to have more
than one mgr module use cherrypy.

I consider this implementation to be a bit messy; I think it'd
be neater if the sub-interpreter was created inside MgrPyModule,
rather than in PyModules::init(), but see the comment in that
function for more details.

Also, due to the lack of documentation on python sub-interpreters,
I'm not completely confident that I'm doing everything correctly.
Notably, there's still calls to PyGILState_Ensure() and
PyEval_ReleaseThread() in PyState.cc, and apparently those two
are either not-compatible with sub-interpreters, or at least need
to be handled delicately.

TL;DR: Everything seems to work for me so far, but should be
considered experimental until further notice.

Signed-off-by: Tim Serong <tserong@suse.com>

### CODE REVIEW

**Comment by tserong** [comment]
Ping @jcsp, @tchaikov, @dmick. I tested this by applying it on top of #14946, and both 'rest' and 'dashboard' modules were able to load and run without stomping on each other.

**Comment by tserong** [comment]
One notable known problem: if a module fails to load because it can't be found (e.g. sys.path is wrong, which shouldn't happen in real life, or a module name is mis-typed in ceph.conf, which might), the python bit of mgr locks up somehow. The loop in PyModules::init() just kinda stops somehow.

**Comment by tserong** [comment]
(That last was just a rebase)

**Comment by tserong** [comment]
OK, I'm confident in this one now.

Each MgrPyModule gets its own python sub-interpreter.  The logger, the ceph_state module and sys.path need to be set up separately for each sub-interpreter, so all that happens in MgrPyModule's constructor (previously this was done on the main interpreter in PyModules::init()).
    
Each sub-interpreter has its own python thread state.  The main interpreter also has a python thread state, but that is almost unused except for during setup and teardown of the whole beast.
    
Some care needs to be taken to ensure that the right thread state is active at the right time; note how the call to handle_pyerror() in PyModules::init() had to be moved inside MgrPyModle::load(). It's not too bad though, just remember that PyModules should be using the main thread state, and MgrPyModule should always be using its sub-interpreter thread state, except for very early in its constructor and very late in its destructor.
    
The ceph_state module (PyState.cc) naturally has no idea what context it's being run in, so uses PyThreadState_Get() when it needs to know the python thread state.
    
I've also added a little Gil class to make it easy to acquire and release the GIL for whichever thread state you're in.

**Comment by tserong** [comment]
Dammit, I just tried using mgr built with this changeset to load the restful module from #14457. It loads fine, but segfaults on shutdown...

**Comment by jcsp** [comment]
This is a tough one to review because the subinterpreter functionality in python itself is so obscure.  It would be nice to spin the Gil class out into a separate commit.

Since PyFinalize allegedly (https://docs.python.org/2.7/c-api/init.html#c.Py_EndInterpreter) will clean up all un-terminated sub interpreters, I wonder if the stuff in ~MgrPyModule could just go away -- it doesn't seem that useful to *maybe* terminate sub interpreters (obviously not your fault, quirk of the framework).  When taking http://tracker.ceph.com/issues/19549 into account, we never truly need to do completely clean teardown, it just needs to be torn down far enough for the process to end cleanly.

**Comment by tserong** [comment]
Ok, I'll split this up into separate commits. And yeah, I wondered the same about getting rid of the attempted termination in ~MgrPyModule, given the quirks. Now that there's two of us, I'll do that as well (but leave some sort of comment).

**Comment by tserong** [comment]
Splitting out the Gil change was a really good idea. Turns out that change by itself exhibited the segfault on shutdown, even without the sub-interpreters, and now I've learned even *more* than I ever wanted to know about embedding Python :-)

Here's what I was missing: not only do we need separate sub-interpreters, we need to manually create an additional python thread state for each OS thread that runs the module's serve function. The PyGILState_*() APIs did that bit automatically, but because I removed those (as they're not supported with sub-interpreters), that extra thread state creation wasn't happening, which resulted in that weird segfault.

**Comment by tserong** [comment]
OK, that's better. No weird segfaults. I think I'll avoid using the word "confident" this time and instead go with cautious optimism (and flashbacks to when I drew [this thing](http://ourobengr.com/ourobengr/) when I first started working on HA).

My one outstanding question now, I think, is: are there any other OS threads that need another manually created python thread state?

**Comment by tserong** [comment]
I've moved the additional thread state creation inside the Gil class, but the caller has to pass new_thread == true when we know we're in a new OS thread . IMO it'd be better to do this automatically, which we can do, if we don't mind using something that is theoretically a private member of PyThreadState (see the comments in Gil.h for details). What do you think @jcsp?

File: src/mgr/Gil.h [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+//
+// See the comment below for when to set new_thread == true
+//
+class Gil {
+public:
+  Gil(PyThreadState *ts, bool new_thread = false)
+  : pThreadState(ts), pNewThreadState(nullptr)
+  {
+    // Acquire the GIL, set the current thread state
+    PyEval_RestoreThread(pThreadState);
```
Comment (chardan) [inline src/mgr/Gil.h]: What happens if pThreadState (ts) is nullptr? The documentation for PyEvalRestoreThread() says that it must not be called on a NULL pointer. Similarly, as I'm reading it, before we do this we have to be sure that we haven't already been called: what happens if this ctor is invoked twice?

File: src/mgr/Gil.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    //
+    //   if (pThreadState->thread_id != PyThread_get_thread_ident()) {
+    //
+    // However, this means we're accessing pThreadState->thread_id, but
+    // the Python C API docs say that "The only public data member is
+    // PyInterpreterState *interp", i.e. doing this would violate
+    // something that's meant to be a black box.
+    //
+    if (new_thread) {
+      pNewThreadState = PyThreadState_New(pThreadState->interp);
```
Comment (chardan) [inline src/mgr/Gil.h]: Does the old state need to be released in some way before re-assigning pNewThreadState?

File: src/mgr/Gil.h (hunk 1 above)
Comment (tserong) [inline src/mgr/Gil.h]: If pThreadState is null it probably dies horribly. I've added an assert (so it'll still die horribly, but on our terms).

As for ensuring we haven't already been called, that's (sadly) up to the caller to be aware of. I've updated the comment to hopefully clarify this a bit more. If you invoke this ctor twice in the same thread it'll deadlock, e.g. do not do this:

````
void func1() {
  Gil gil(pMyThreadState);
  // Do some stuff that needs the GIL here

  func2();  // this will deadlock
}
void func2() {
  Gil gil(pMyThreadState);
  // Do some stuff that needs the GIL
}
````

You can avoid this deadlock like so:

````
void func1() {
  {  // extra scope
    Gil gil(pMyThreadState);
    // Do some stuff that needs the GIL here
  }

  func2();  // this won't deadlock
}
````

In summary, you have to be just as careful using this class as you need to be when using PyEvalRestoreThread(). AFAICT there's no public API for figuring out whether or not the GIL is already held.

File: src/mgr/Gil.h (hunk 2 above)
Comment (tserong) [inline src/mgr/Gil.h]: No. At this point, pNewThreadState is null, and PyThreadState_New creates a new PyThreadState attached to the same interpreter that the existing pThreadState is using. It's fine for the two PyThreadState objects to exist at the same time; only one is ever active (the PyThreadState_Swap() calls enable activate or the other).

File: src/mgr/Gil.h [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+// know you need the GIL in.
+//
+// See the comment below for when to set new_thread == true
+//
+class Gil {
+public:
+  Gil(PyThreadState *ts, bool new_thread = false)
+  : pThreadState(ts), pNewThreadState(nullptr)
+  {
+    assert(pThreadState != nullptr);
```
Comment (chardan) [inline src/mgr/Gil.h]: Consider-also throwing an exception instead.

File: src/mgr/Gil.h (hunk 2 above)
Comment (chardan) [inline src/mgr/Gil.h]: Given that we're now sure pNewThreadState is nullptr, this sounds good. Thanks!

File: src/mgr/Gil.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      PyThreadState_Delete(pNewThreadState);
+    }
+    // Release the GIL, reset the thread state to NULL
+    PyEval_SaveThread();
+    dout(20) << "GIL released for thread state " << pThreadState << dendl;
+  }
+
+private:
+  PyThreadState *pThreadState;
+  PyThreadState *pNewThreadState;
```
Comment (chardan) [inline src/mgr/Gil.h]: I think initializing these in the ctor's initializer-list is fine, but since we're using C++ it's also possible to write:
PyThreadState *pNewThreadState = nullptr;
(And/or use std::unique_ptr<>.)
...I'm fine either way, but figured I should point it out.

File: src/mgr/MgrPyModule.cc [hunk 5]
```diff
@@ -46,32 +48,122 @@ std::string handle_pyerror()
...
+namespace {
+  PyObject* log_write(PyObject*, PyObject* args) {
+    char* m = nullptr;
+    if (PyArg_ParseTuple(args, "s", &m)) {
+      auto len = strlen(m);
+      if (len && m[len-1] == '\n') {
+	m[len-1] = '\0';
+      }
+      dout(4) << m << dendl;
+    }
```
Comment (chardan) [inline src/mgr/MgrPyModule.cc]: If PyArg_ParseTuple has returned false, we're swallowing the returned error here. Certainly, fine if we didn't want to take any action, but is this intentional?

File: src/mgr/MgrPyModule.cc [hunk 6]
```diff
@@ -46,32 +48,122 @@ std::string handle_pyerror()
...
+#define dout_prefix *_dout << "mgr[py] "
+
+namespace {
+  PyObject* log_write(PyObject*, PyObject* args) {
+    char* m = nullptr;
+    if (PyArg_ParseTuple(args, "s", &m)) {
+      auto len = strlen(m);
+      if (len && m[len-1] == '\n') {
+	m[len-1] = '\0';
+      }
```
Comment (chardan) [inline src/mgr/MgrPyModule.cc]: If len is one, this assumes not-an-error (m[len-1] won't explode because len > 0 on account of the first clause). I don't see in the Python documentation whether or not that's possible, but if it is it might be worth checking for.

File: src/mgr/MgrPyModule.cc [hunk 7]
```diff
@@ -46,32 +48,122 @@ std::string handle_pyerror()
...
-{}
-
-MgrPyModule::~MgrPyModule()
+    pClassInstance(nullptr),
+    pMainThreadState(main_ts_),
+    pMyThreadState(nullptr)
 {
-  PyGILState_STATE gstate;
-  gstate = PyGILState_Ensure();
+  assert(pMainThreadState != nullptr);
```
Comment (chardan) [inline src/mgr/MgrPyModule.cc]: Consider throwing an exception.

File: src/mgr/MgrPyModule.cc [hunk 8]
```diff
@@ -46,32 +48,122 @@ std::string handle_pyerror()
...
+
+  Gil gil(pMainThreadState);
+
+  pMyThreadState = Py_NewInterpreter();
+  if (pMyThreadState == nullptr) {
+    derr << "Failed to create python sub-interpreter for '" << module_name << '"' << dendl;
+  } else {
```
Comment (chardan) [inline src/mgr/MgrPyModule.cc]: If pMyThreadState == nullptr, the program continues, leaving the object constructed. This seems like a critical failure to me? If so, I think it would be better to throw an exception than to leave the object in an invalid state.

File: src/mgr/MgrPyModule.cc [hunk 9]
```diff
@@ -46,32 +48,122 @@ std::string handle_pyerror()
...
+    const char *argv[] = {"ceph-mgr"};
+    PySys_SetArgv(1, (char**)argv);
+
+    if (g_conf->daemonize) {
+      auto py_logger = Py_InitModule("ceph_logger", log_methods);
+#if PY_MAJOR_VERSION >= 3
+      PySys_SetObject("stderr", py_logger);
+      PySys_SetObject("stdout", py_logger);
+#else
+      PySys_SetObject(const_cast<char*>("stderr"), py_logger);
```
Comment (chardan) [inline src/mgr/MgrPyModule.cc]: Is the first ifdef needed? (eg. will the version here with the cast always work anyway?)

File: src/mgr/MgrPyModule.cc [hunk 10]
```diff
@@ -46,32 +48,122 @@ std::string handle_pyerror()
...
+      PySys_SetObject(const_cast<char*>("stdout"), py_logger);
+#endif
+    }
+    // Populate python namespace with callable hooks
+    Py_InitModule("ceph_state", CephStateMethods);
```
Comment (chardan) [inline src/mgr/MgrPyModule.cc]: Should the return value of this function be checked?

File: src/mgr/MgrPyModule.cc [hunk 11]
```diff
@@ -46,32 +48,122 @@ std::string handle_pyerror()
...
 
-  Py_XDECREF(pClassInstance);
+    PySys_SetPath((char*)(sys_path.c_str()));
```
Comment (chardan) [inline src/mgr/MgrPyModule.cc]: Prefer C++ style casts. Even though they're annoying to type. :-)

File: src/mgr/MgrPyModule.cc [hunk 12]
```diff
@@ -46,32 +48,122 @@ std::string handle_pyerror()
...
 }
 
 int MgrPyModule::load()
 {
+  if (pMyThreadState == nullptr) {
+    derr << "No python sub-interpreter exists for module '" << module_name << "'" << dendl;
+    return -EINVAL;
+  }
+
+  Gil gil(pMyThreadState);
```
Comment (chardan) [inline src/mgr/MgrPyModule.cc]: I'm starting to wonder if it's worth calling type Gil something like "Gil_scope_guard"?

File: src/mgr/MgrPyModule.cc [hunk 13]
```diff
@@ -46,32 +48,122 @@ std::string handle_pyerror()
...
+
   // Load the module
   PyObject *pName = PyString_FromString(module_name.c_str());
```
Comment (chardan) [inline src/mgr/MgrPyModule.cc]: We're guaranteed that module_name.c_str() will not be NULL, but PyString_FromString() is going to make a copy, which could fail. We may want to check pName for nullptr.

File: src/mgr/MgrPyModule.cc [hunk 14]
```diff
@@ -201,15 +287,12 @@ void MgrPyModule::notify_clog(const LogEntry &log_entry)
...
-  PyGILState_Release(gstate);
 }
 
 int MgrPyModule::load_commands()
 {
-  PyGILState_STATE gstate;
-  gstate = PyGILState_Ensure();
-
+  // Don't need a Gil here -- this is called from MgrPyModule::load(),
+  // which already has one.
```
Comment (chardan) [inline src/mgr/MgrPyModule.cc]: It is really too bad that there's no way to detect if two are the same! :-\

File: src/mgr/PyModules.cc [hunk 15]
```diff
@@ -366,56 +339,39 @@ int PyModules::init()
...
     if (r != 0) {
+      // Don't use handle_pyerror() here; we don't have the GIL
+      // or the right thread state (this is deliberate).
       derr << "Error loading module '" << module_name << "': "
         << cpp_strerror(r) << dendl;
-      derr << handle_pyerror() << dendl;
       // Don't drop out here, load the other modules
     } else {
       // Success!
       modules[module_name] = std::move(mod);
```
Comment (chardan) [inline src/mgr/PyModules.cc]: std::move() is fine here, but since it's a unique_ptr<> you can also just assign it. (No harm done.)

File: src/mgr/Gil.h [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+    if (new_thread) {
+      pNewThreadState = PyThreadState_New(pThreadState->interp);
+      PyThreadState_Swap(pNewThreadState);
+      dout(20) << "Switched to new thread state " << pNewThreadState << dendl;
+    }
+  }
+
+  ~Gil()
+  {
+    if (pNewThreadState != nullptr) {
```
Comment (chardan) [inline src/mgr/Gil.h]: What you've written is fine, but another option to consider might be a std::unique_ptr<> with a deleter, so this explicit checking wouldn't be necessary.

File: src/mgr/MgrPyModule.cc [hunk 17]
```diff
@@ -49,29 +50,32 @@ std::string handle_pyerror()
...
 
-MgrPyModule::MgrPyModule(const std::string &module_name_)
+MgrPyModule::MgrPyModule(const std::string &module_name_, PyThreadState *main_ts_)
   : module_name(module_name_),
-    pClassInstance(nullptr)
-{}
+    pClassInstance(nullptr),
+    pMainThreadState(main_ts_)
+{
+  assert(pMainThreadState != nullptr);
```
Comment (chardan) [inline src/mgr/MgrPyModule.cc]: (As above-- considering throwing.)

File: src/mgr/PyState.cc [hunk 18]
```diff
@@ -51,28 +54,28 @@ class MonCommandCompletion : public Context
...
-    auto pyOutS = PyString_FromString(outs.c_str());
-    auto args = PyTuple_Pack(3, pyR, pyOutBl, pyOutS);
-    Py_DECREF(pyR);
-    Py_DECREF(pyOutBl);
-    Py_DECREF(pyOutS);
-
-    auto rtn = PyObject_CallObject(set_fn, args);
-    if (rtn != nullptr) {
-      Py_DECREF(rtn);
+    dout(10) << "MonCommandCompletion::finish()" << dendl;
```
Comment (chardan) [inline src/mgr/PyState.cc]: I'm a bit curious: couldn't we use the function's scope if we were willing to put this dout() line just before the Gil construction? Is there ya reason we don't? (I'm not fully initiated into the Great Mystery of dout().)

File: src/mgr/PyState.cc [hunk 19]
```diff
@@ -51,28 +54,28 @@ class MonCommandCompletion : public Context
...
+      if (rtn != nullptr) {
+	Py_DECREF(rtn);
+      }
+      Py_DECREF(args);
     }
-    Py_DECREF(args);
-
-    PyGILState_Release(gstate);
-
     global_handle->notify_all("command", tag);
```
Comment (chardan) [inline src/mgr/PyState.cc]: ...ah-hah, this may be the reason for the extra scope?

File: src/mgr/PyState.cc [hunk 20]
```diff
@@ -51,28 +54,28 @@ class MonCommandCompletion : public Context
...
-      Py_DECREF(rtn);
+    dout(10) << "MonCommandCompletion::finish()" << dendl;
+    {
+      // Scoped so the Gil is released before calling notify_all()
+      Gil gil(pThreadState);
+
+      auto set_fn = PyObject_GetAttrString(python_completion, "complete");
+      assert(set_fn != nullptr);
+
+      auto pyR = PyInt_FromLong(r);
```
Comment (chardan) [inline src/mgr/PyState.cc]: Some of these calls can fail. Granted, the usual reason will be memory errors which can be hard to handle, but I figured I'd point it out.

File: src/mgr/Gil.h (hunk 1 above)
Comment (chardan) [inline src/mgr/Gil.h]: +1

File: src/mgr/Gil.h [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+//
+// Do *not* nest these, as a second GIL acquire will deadlock (see
+// https://docs.python.org/2/c-api/init.html#c.PyEval_RestoreThread)
+//
+// If in doubt, explicitly put a scope around the block of code you
+// know you need the GIL in.
+//
+// See the comment below for when to set new_thread == true
+//
+class Gil {
```
Comment (chardan) [inline src/mgr/Gil.h]: One more thing I just thought of. It might also be worth disabling the copy ctor and assignment operators, eg.
Gil(const Gil&) = delete;
Gil& operator=(const Gil&) = delete;

File: src/mgr/MgrPyModule.cc (hunk 8 above)
Comment (tserong) [inline src/mgr/MgrPyModule.cc]: I wondered about that, but couldn't see any other exceptions being thrown anywhere in the mgr codebase, so decided not to go there.

The way these things get constructed (in PyModules::init()), the next thing that's called is MgrPyModule::load() which will return with an error if the object is in that invalid state. So it's handled, if not completely clean.

File: src/mgr/PyState.cc (hunk 19 above)
Comment (tserong) [inline src/mgr/PyState.cc]: Yep.

File: src/mgr/MgrPyModule.cc (hunk 12 above)
Comment (tserong) [inline src/mgr/MgrPyModule.cc]: That's probably a good idea - it'd make it more obvious to the reader WTF it was for.

File: src/mgr/Gil.h (hunk 16 above)
Comment (tserong) [inline src/mgr/Gil.h]: I think I'd like to stick with the explicit check, just because what I'm doing here is small, self contained and (hopefully) pretty obvious.

**Comment by tserong** [comment]
Thanks for the detailed review @chardan. I've done a couple of tiny bits of cleanup (3e0fb3e). Your remaining comments that I haven't either made a change for, or responded to, apply to existing code (which I might have moved around for this PR). I'd like to hang back on doing anything about those until after @tchaikov and/or @jcsp are able to do some review of this PR as well.

**Comment by chardan** [comment]
You're very welcome! Nice work on this change. I didn't really see any serious problems, so just about all of my suggestions are... suggestions. :-) Nice job, @tserong!

**Comment by tserong** [comment]
Rebased & squashed some cleanup commits.

**Comment by tserong** [comment]
Now that #14946 is merged, this kinda needs to go in too in case someone enables the rest and dashboard modules at the same time (if both are enabled, neither will work, because of the two cherrypy's stomping on each other).

**Comment by tserong** [comment]
retest this please

**Comment by yuriw** [comment]
ready for merge

http://pulpito.ceph.com/yuriw-2017-05-26_19:34:44-rados-wip-yuri-testing_2017_5_27-distro-basic-smithi/

http://pulpito.ceph.com/yuriw-2017-05-29_22:24:45-rados-wip-yuri-testing_2017_5_27---basic-smithi/
